File: jcrm-ui/src/app/resources/user-resource.ts
import {MinimalUserResource} from '@app/resource/minimal-user-resource';
import {PhoneNumberResource} from '@app/resource/phone-number-resource';
import {User} from '@app/entity/user';
import {MessengerResource} from '@app/resource/messenger-resource';

export class UserResource extends MinimalUserResource {

    public static toUser(resource: UserResource): User {
        return new User(
            resource.id,
            resource.email,
            resource.firstName,
            resource.secondName,
            resource.thirdName,
            resource.birthday ? new Date(resource.birthday) : null,
            PhoneNumberResource.toPhoneNumbers(resource.phoneNumbers),
            MessengerResource.toMessengers(resource.messengers),
            resource.password,
            resource.created ? new Date(resource.created) : null,
            resource.modified ? new Date(resource.modified) : null
        );
    }

    constructor(
        id: number | null,
        email: string | null,
        firstName: string | null,
        secondName: string | null,
        thirdName: string | null,
        birthday: string | null,
        public phoneNumbers: PhoneNumberResource[] | null,
        public messengers: MessengerResource[] | null,
        public password: string | null,
        public created: number | null,
        public modified: number | null,
    ) {
        super(id, email, firstName, secondName, thirdName, birthday);
    }
}

File: jcrm_ui_flutter/lib/src/jcrm/ui/flutter/resource/user_resource.dart
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/json_resource.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/minimal_user_resource.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/phone_number_resource.dart';
import 'package:jcrm_ui_flutter/src/jcrm/ui/flutter/resource/messenger_resource.dart';

class UserResource extends MinimalUserResource {
  final List<PhoneNumberResource> phoneNumbers;
  final List<MessengerResource> messengers;
  final int created;
  final int modified;

  UserResource(int id, String email, String firstName, String secondName, String thirdName, String birthday,
      this.phoneNumbers, this.messengers, this.created, this.modified)
      : super(id, email, firstName, secondName, thirdName, birthday);

  UserResource.fromJson(Map<String, dynamic> json)
      : phoneNumbers = PhoneNumberResource.fromJsonList(json['phoneNumbers']),
        messengers = MessengerResource.fromJsonList(json['messengers']),
        created = json['created'],
        modified = json['modified'],
        super.fromJson(json);

  @override
  void buildJson(Map<String, dynamic> json) {
    super.buildJson(json);
    json['phoneNumbers'] = JsonResource.toJsonList(phoneNumbers);
    json['messengers'] = JsonResource.toJsonList(messengers);
    json['created'] = created;
    json['modified'] = modified;
  }
}
